### PR TITLE
Support for skin tones in the picker

### LINKED ISSUE #80: Implement long press on emoji to switch skin tone
Can be implemented once #77 is done.

### PR DESCRIPTION
**Pull request by rubengees**
This would fix #80.

### Contents of this PR

A popup when clicked on an Emoji to show the available skin colors, similar to WhatsApp. It looks like this:

device-2017-02-18-001345

#### Code changes

- A new `OnEmojiLongClickedListener`, which is delegated through `Adapters` and `Views`, similar to the existing `OnEmojiClickedListener`. It is used to detect long clicks and show the popup.
- An additional method for finding skin toned Emojis, given an `Emoji` instance, for the `EmojiTree` and `EmojiManager`.
- A new class `EmojiSkinTonePopup`, which is used in the `EmojiPopup` and displays the popup.
- Small cleanups here and there.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -54,6 +54,7 @@ ext {
     supportV4CoreUi = "com.android.support:support-core-ui:$supportLibraryVersion"
     supportV4Compat = "com.android.support:support-compat:$supportLibraryVersion"
     supportRecyclerView = "com.android.support:recyclerview-v7:${supportLibraryVersion}"
+    supportCardView = "com.android.support:cardview-v7:${supportLibraryVersion}"
```
Comment (rubengees) [inline build.gradle]: Is it okay to add the `cardview` dependency? I found no other good looking way to add elevation to the new popup, while maintaining backwards compability.

Comment (vanniktech) [inline build.gradle]: I'd prefer if we would not use it. There must be a way to add elevation to it without CardView.

File: emoji/src/main/java/com/vanniktech/emoji/EmojiArrayAdapter.java [hunk 2]
```diff
@@ -10,20 +10,27 @@
...
+import com.vanniktech.emoji.listeners.OnEmojiLongClickedListener;
+
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.List;
 
 import static com.vanniktech.emoji.Utils.checkNotNull;
 
 final class EmojiArrayAdapter extends ArrayAdapter<Emoji> {
+  private static final String EMOJI_NULL_ERROR = "emoji == null";
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiArrayAdapter.java]: remove this, we'll adjust the PMD / Checkstyle config

File: emoji/src/main/java/com/vanniktech/emoji/EmojiArrayAdapter.java [hunk 3]
```diff
@@ -55,6 +62,23 @@ public View getView(final int position, final View convertView, @NonNull final V
       }
     });
 
+    if (!EmojiManager.getInstance().findSkinTonedEmojis(checkNotNull(getItem(position), EMOJI_NULL_ERROR)).isEmpty()) {
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiArrayAdapter.java]: you can extract the variable `checkNotNull(getItem(position), EMOJI_NULL_ERROR)` here

File: emoji/src/main/java/com/vanniktech/emoji/EmojiArrayAdapter.java [hunk 4]
```diff
@@ -55,6 +62,23 @@ public View getView(final int position, final View convertView, @NonNull final V
...
+      image.setOnLongClickListener(new View.OnLongClickListener() {
+          @Override
+          public boolean onLongClick(final View v) {
+            if (longListener == null) {
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiArrayAdapter.java]: invert if statement and remove the else block

File: emoji/src/main/java/com/vanniktech/emoji/EmojiGridView.java [hunk 5]
```diff
@@ -24,8 +26,11 @@
     setClipToPadding(false);
   }
 
-  public EmojiGridView init(@Nullable final OnEmojiClickedListener onEmojiClickedListener, @NonNull final EmojiCategory category) {
-    emojiArrayAdapter = new EmojiArrayAdapter(getContext(), category.getEmojis(), onEmojiClickedListener);
+  public EmojiGridView init(@Nullable final OnEmojiClickedListener onEmojiClickedListener,
+                            @Nullable final OnEmojiLongClickedListener onEmojiLongClickedListener,
+                            @NonNull final EmojiCategory category) {
+    emojiArrayAdapter = new EmojiArrayAdapter(getContext(), category.getEmojis(),
+            onEmojiClickedListener, onEmojiLongClickedListener);
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiGridView.java]: put it in one line

File: emoji/src/main/java/com/vanniktech/emoji/EmojiGridView.java [hunk 6]
```diff
@@ -4,8 +4,10 @@
 import android.support.annotation.NonNull;
 import android.support.annotation.Nullable;
 import android.widget.GridView;
+
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiGridView.java]: nit: remove

File: emoji/src/main/java/com/vanniktech/emoji/EmojiArrayAdapter.java [hunk 7]
```diff
@@ -10,20 +10,27 @@
 import android.widget.ImageView;
 import com.vanniktech.emoji.emoji.Emoji;
 import com.vanniktech.emoji.listeners.OnEmojiClickedListener;
+import com.vanniktech.emoji.listeners.OnEmojiLongClickedListener;
+
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiArrayAdapter.java]: nit: remove

File: emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java [hunk 8]
```diff
@@ -2,10 +2,13 @@
 
 import android.support.annotation.NonNull;
 import android.support.annotation.Nullable;
+
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java]: nit: remove

File: emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java [hunk 9]
```diff
@@ -2,10 +2,13 @@
...
 import com.vanniktech.emoji.emoji.Emoji;
 import com.vanniktech.emoji.emoji.EmojiCategory;
 import com.vanniktech.emoji.emoji.EmojiTree;
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java]: nit: remove

File: emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java [hunk 10]
```diff
@@ -2,10 +2,13 @@
...
 
+import java.util.List;
+
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java]: nit: remove

File: emoji/src/main/java/com/vanniktech/emoji/EmojiPagerAdapter.java [hunk 11]
```diff
@@ -29,7 +34,8 @@
 
       recentEmojiGridView = (RecentEmojiGridView) newView;
     } else {
-      newView = new EmojiGridView(pager.getContext()).init(listener, EmojiManager.getInstance().getCategories()[position - 1]);
+      newView = new EmojiGridView(pager.getContext()).init(listener, longListener,
+              EmojiManager.getInstance().getCategories()[position - 1]);
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiPagerAdapter.java]: can be one line

File: emoji/src/main/java/com/vanniktech/emoji/EmojiPopup.java [hunk 12]
```diff
@@ -16,9 +16,11 @@
 import android.view.WindowManager;
 import android.view.inputmethod.InputMethodManager;
 import android.widget.PopupWindow;
+
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiPopup.java]: nit: remove

File: emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+import android.widget.ImageView;
+import android.widget.LinearLayout;
+import android.widget.PopupWindow;
+
+import com.vanniktech.emoji.emoji.Emoji;
+import com.vanniktech.emoji.listeners.OnEmojiClickedListener;
+
+import java.util.List;
+
+class EmojiSkinTonePopup {
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java]: final

File: emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+    popupWindow = new PopupWindow(content,
+            WindowManager.LayoutParams.WRAP_CONTENT,
+            WindowManager.LayoutParams.WRAP_CONTENT
+    );
+
+    final List<Emoji> skinTonedEmojis = EmojiManager.getInstance().findSkinTonedEmojis(emoji);
+    skinTonedEmojis.add(0, emoji);
+
+    for (final Emoji tonedEmoji : skinTonedEmojis) {
+      final ImageView emojiImage = (ImageView) LayoutInflater.from(clickedImage.getContext())
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java]: `LayoutInflater.from(clickedImage.getContext())` into variable and before the loop so we don't look it up in a loop

File: emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+              .inflate(R.layout.emoji_item, imageContainer, false);
+      final ViewGroup.MarginLayoutParams layoutParams = (ViewGroup.MarginLayoutParams) emojiImage.getLayoutParams();
+      final int margin = (int) Utils.dpToPx(clickedImage.getContext(), MARGIN);
+
+      // Use the same size for Emojis as the picker
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java]: http://nedbatchelder.com/blog/201401/comments_should_be_sentences.html

File: emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+    );
+
+    final List<Emoji> skinTonedEmojis = EmojiManager.getInstance().findSkinTonedEmojis(emoji);
+    skinTonedEmojis.add(0, emoji);
+
+    for (final Emoji tonedEmoji : skinTonedEmojis) {
+      final ImageView emojiImage = (ImageView) LayoutInflater.from(clickedImage.getContext())
+              .inflate(R.layout.emoji_item, imageContainer, false);
+      final ViewGroup.MarginLayoutParams layoutParams = (ViewGroup.MarginLayoutParams) emojiImage.getLayoutParams();
+      final int margin = (int) Utils.dpToPx(clickedImage.getContext(), MARGIN);
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java]: cast seems redundant

File: emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+          if (listener != null) {
+            listener.onEmojiClicked(tonedEmoji);
+          }
+        }
+      });
+
+      imageContainer.addView(emojiImage);
+    }
+
+    content.measure(View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED),
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java]: static import `makeMeasureSpec`

File: emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+            View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED));
+
+    popupWindow.setOutsideTouchable(true);
+    popupWindow.setBackgroundDrawable(new BitmapDrawable(clickedImage.getContext().getResources(), (Bitmap) null));
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java]: What is this line doing?

File: emoji/src/main/java/com/vanniktech/emoji/RecentEmojiGridView.java [hunk 19]
```diff
@@ -18,7 +18,7 @@ public RecentEmojiGridView init(@Nullable final OnEmojiClickedListener onEmojiCl
     recentEmojis = recentEmoji;
 
     final Collection<Emoji> emojis = recentEmojis.getRecentEmojis();
-    emojiArrayAdapter = new EmojiArrayAdapter(getContext(), emojis.toArray(new Emoji[emojis.size()]), onEmojiClickedListener);
+    emojiArrayAdapter = new EmojiArrayAdapter(getContext(), emojis.toArray(new Emoji[emojis.size()]), onEmojiClickedListener, null);
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/RecentEmojiGridView.java]: We should also here add the ability to long press and select a different skin tone

File: emoji/src/main/java/com/vanniktech/emoji/Utils.java [hunk 20]
```diff
@@ -27,6 +28,10 @@
     return reference;
   }
 
+  public static float dpToPx(final Context context, final float dp) {
+    return dp * context.getResources().getDisplayMetrics().density;
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/Utils.java]: do the cast here and return int

File: emoji/src/main/java/com/vanniktech/emoji/emoji/EmojiTree.java [hunk 21]
```diff
@@ -5,12 +5,19 @@
 import android.support.annotation.RestrictTo;
 import android.support.v4.util.SparseArrayCompat;
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/emoji/EmojiTree.java]: nit: remove line

File: emoji/src/main/java/com/vanniktech/emoji/emoji/EmojiTree.java [hunk 22]
```diff
@@ -43,6 +49,36 @@ public void add(@NonNull final Emoji emoji) {
     return result;
   }
 
+  public List<Emoji> findSkinTonedEmojis(@NonNull final Emoji emoji){
+    EmojiNode current = root;
+
+    for (int i = 0; i < emoji.getUnicode().length(); i++){
+      current = current.getChild(emoji.getUnicode().charAt(i));
+
+      if (current == null){
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/emoji/EmojiTree.java]: nit: space between `){`

File: emoji/src/main/java/com/vanniktech/emoji/emoji/EmojiTree.java [hunk 23]
```diff
@@ -43,6 +49,36 @@ public void add(@NonNull final Emoji emoji) {
     return result;
   }
 
+  public List<Emoji> findSkinTonedEmojis(@NonNull final Emoji emoji){
+    EmojiNode current = root;
+
+    for (int i = 0; i < emoji.getUnicode().length(); i++){
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/emoji/EmojiTree.java]: extract `emoji.getUnicode()`

File: emoji/src/main/java/com/vanniktech/emoji/emoji/EmojiTree.java [hunk 24]
```diff
@@ -43,6 +49,36 @@ public void add(@NonNull final Emoji emoji) {
...
+      current = current.getChild(emoji.getUnicode().charAt(i));
+
+      if (current == null){
+        throw new IllegalArgumentException("Invalid emoji");
+      }
+    }
+
+    current = current.getChild(SKIN_TONE_PART);
+
+    if (current == null) {
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/emoji/EmojiTree.java]: can you reverse this?

```java
if (current != null) {
  // All of the logic.
  return result;
}

return Collections.emptyList();
```

File: emoji/src/main/java/com/vanniktech/emoji/listeners/OnEmojiLongClickedListener.java [hunk 25]
```diff
@@ -0,0 +1,11 @@
+package com.vanniktech.emoji.listeners;
+
+import android.support.annotation.RestrictTo;
+import android.view.View;
+
+import com.vanniktech.emoji.emoji.Emoji;
+
+@RestrictTo(RestrictTo.Scope.LIBRARY)
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/listeners/OnEmojiLongClickedListener.java]: same line as interface

File: emoji/src/main/java/com/vanniktech/emoji/listeners/OnEmojiLongClickedListener.java [hunk 26]
```diff
@@ -0,0 +1,11 @@
...
+public interface OnEmojiLongClickedListener {
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/listeners/OnEmojiLongClickedListener.java]: make it package private and put it into the root if we do not expose this (to be consistent)

File: emoji/src/main/java/com/vanniktech/emoji/listeners/OnEmojiLongClickedListener.java [hunk 27]
```diff
@@ -0,0 +1,11 @@
+package com.vanniktech.emoji.listeners;
+
+import android.support.annotation.RestrictTo;
+import android.view.View;
+
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/listeners/OnEmojiLongClickedListener.java]: nit: remove

**Comment by rubengees** [comment]
So I investigated elevation further and I see 4 ways:

- Use native elevation: Looks bad on pre-lollipop
- Use a custom `layer-list`: No shadows at top, left and right and the shadow has no gradient

layer_list

- Use the `dialog_holo_light_frame` drawable: has too much shadow and does not look really good

holo

- Use `cardview`: Increases method count and general size

I am still pro-cardview. I looked into it, when using `Proguard` with optimization, it adds only 33 methods and some resources. Thats not too bad IMHO. Plus nearly every App uses `appcompat-v7` and I can imagine many use `cardview` as well. It is also the "official way" for doing this.

What do you think?

I will work on the other things now.

File: emoji/src/main/java/com/vanniktech/emoji/EmojiArrayAdapter.java (hunk 7 above)
Comment (rubengees) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiArrayAdapter.java]: Android Studio is always trying to change it back, is it really necessary?
Out of curiosity, why does your code style differ so much from the Android Studio defaults? I have hard times getting Android Studio to not reformat 😁

File: emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java (hunk 16 above)
Comment (rubengees) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java]: `dpToPx` returns `float`, but we need `int` here.
The code is from [this](http://stackoverflow.com/a/12147550/4279995) StackOverflow answer.

File: emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java (hunk 18 above)
Comment (rubengees) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java]: I stole it from [here](https://github.com/vanniktech/Emoji/blob/master/emoji/src/main/java/com/vanniktech/emoji/EmojiPopup.java#L100), I don't know if it is really required though (No visible change with or without).

File: emoji/src/main/java/com/vanniktech/emoji/Utils.java (hunk 20 above)
Comment (rubengees) [inline emoji/src/main/java/com/vanniktech/emoji/Utils.java]: Ah good point, did not read until here, before I answered above.

File: emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java (hunk 10 above)
Comment (rubengees) [inline emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java]: `checkstyle` complains when removing the space between normal and static imports, so I will keep it as it is.

**Comment by rubengees** [comment]
Everything else is done!

File: emoji/src/main/java/com/vanniktech/emoji/emoji/EmojiTree.java [hunk 28]
```diff
@@ -43,6 +49,47 @@ public void add(@NonNull final Emoji emoji) {
     return result;
   }
 
+  public List<Emoji> findSkinTonedEmojis(@NonNull final Emoji emoji){
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/emoji/EmojiTree.java]: Can we add tests for this class?

File: emoji/src/main/java/com/vanniktech/emoji/emoji/EmojiTree.java [hunk 29]
```diff
@@ -43,6 +49,47 @@ public void add(@NonNull final Emoji emoji) {
...
+        }
+      }
+
+      return result;
+    }
+
+    return Collections.emptyList();
+  }
+
+  public Emoji findNonSkinTonedEmoji(@NonNull final Emoji emoji){
```
Comment (vanniktech) [inline emoji/src/main/java/com/vanniktech/emoji/emoji/EmojiTree.java]: Can we add tests for this class?

**Comment by rubengees** [comment]
- No there is no reason, I will change that.
- Jup, I will try to replicate that.
- I noticed that not all Emojis have the skin tone modifier at the end, which I assumed :/. I'm thinking of a way to handle this.
- Thats quite hard. We would have to look up for every emoji if another skin tone is selected and then find the appropriate one
- Hmm, should be possible.

File: emoji/src/main/java/com/vanniktech/emoji/emoji/EmojiTree.java (hunk 28 above)
Comment (rubengees) [inline emoji/src/main/java/com/vanniktech/emoji/emoji/EmojiTree.java]: Will do once all features are in place~

**Comment by vanniktech** [comment]
@rubengees I just saw that Whatsapp displays every skin tone in the recent one so we can leave the behaviour and I'll take it off the list. About the 4th point, if you want I can also do this once this PR is merged.

**Comment by rubengees** [comment]
I have spent some time on the skin tone problem and pushed a new approach:

Every `Emoji` now has references to its variants (skin tones, maybe more in the future) and the variants back to the base `Emoji` without modifications. This also greatly simplifies the `EmojiTree` and leaves the heavy work to the `generator`, which was adjusted for this. One drawback is the slightly increased memory consumption (Due to the additional references), but I think it's not too bad.

I also made the popup close on selection.
About the 2. point: I can't find a drawable like that one WhatsApp and others are using. Do you have an idea where to look?

**Comment by vanniktech** [comment]
One more thing, I think that the indent level is off as it shows that it has changed.

**Comment by rubengees** [comment]
I have renamed the `SquareImageView` to `EmojiImageView` and extended it to draw the indicator rectangle. I have also changed some of the size and margin values to make it look good.
Moreover I have added some tests and cleaned up the code a bit.

Please have a look!

**Comment by rubengees** [comment]
Regarding the indentation: The changes you are seeing are from the variants. I though indenting them a bit is a good idea, to make it visible that they are no base emojis, added directly to the tree (Take a closer look at the parentheses: They are passed to the constructor of their respective base emoji).

**Comment by rubengees** [comment]
Formatting adjusted!

**Comment by vanniktech** [comment]
Thank you once again 👍

**Pull request merged**